### PR TITLE
add frogger game

### PR DESCRIPTION
**Pull request by YuliiaDikun**
### Classic Frogger Game

[Demo](https://yuliiadikun.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/YuliiaDikun/frontend-nanodegree-arcade-game/blob/master/js/app.js)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by YuliiaDikun** [comment]
I have completed self-checks and fixed the code accordingly. Please, look at it. Thanks for your time.

**Comment by OleksiyRudenko** [comment]
Also please before submitting your next task make your code editor do indents with 2 spaces, not tabs.
Tabs expand in different ways on different environments. 8 spaces indentation at times makes reading and following code structure a bit painful.

Do not do this for this task as it will mark the entire code as changed.

File: submissions/dikun/frogger-game/app.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
+"use srtict";
+
+const cell = {
+	width: 100,
+	height: 85
+};
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:6]: By convention, constants are named all caps and snake-case (e.g. `CELL`, `GAME_FIELD`). This makes them outstand in the code and let the reader instantly know they are constants and not just global variables.

File: submissions/dikun/frogger-game/app.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	max: 250 
+}; 
+
+const sprite = {
+	enemy: 'images/enemy-bug.png',
+	player: 'images/char-boy.png'
+};
+
+const speed = () => enemySpeed.min + Math.floor(Math.random() * enemySpeed.max);
+const crash = 70;
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:31]: As a peer developer I want to understand what does `crash` mean and how do I use it. Is this power of crash, overlap by one of axis, overlap amount or anything else?

File: submissions/dikun/frogger-game/app.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+	this.checkColission = function () {
+		if(this.player.x + crash > this.x &&
+			this.player.x < this.x + crash &&
+			this.player.y + crash > this.y &&
+			this.player.y < this.y + crash) {
+			this.player.x = playerStart.x;
+			this.player.y = playerStart.y;
+		}
+	};
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:49]: This approach defines a **copy** of function per every instance of class.
We do not want this. One of the crucial points of prototypical inheritance model is to have a single copy of method per class, stored in its prototype object.
Check the original project code for a pattern to follow.

File: submissions/dikun/frogger-game/app.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			alert('You win!');	
+		}			
+	};
+};
+
+Function.prototype.construct = function (position) {
+	let gameEnteties = Object.create(this.prototype);
+	this.apply(gameEnteties, position);
+	return gameEnteties;
+};
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:102]: We do not want to play around with JS built in classes/objects.
If you want a base class, you need a class function like `Player` or `Enemy` and then you bind `Player` and `Enemy` classes with this base class by `Object.create`.

File: submissions/dikun/frogger-game/app.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+const playerPosition = [playerStart.x, playerStart.y];
+const player = Player.construct(playerPosition);
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:105]: Adjust the approach.
An instance would be created with the `new` keyword.

File: submissions/dikun/frogger-game/app.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+const allEnemies = [cell.height - border, cell.height*2 - border, cell.height*3 - border].map((value) => {
+	const enemyPosition = [gameField.left, value, speed, player];
+	return Enemy.construct(enemyPosition);
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:109]: The composition of an array to the right doesn't look like a position altogether.
Also class functions are quite normal functions and can have normal named parameters.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@YuliiaDikun a pretty decent job!
But something went sideways with class methods declaration and building a base class.

Feel free adding changes incrementally and asking questions.

I also wonder where did you pick this class methods definition pattern?

File: submissions/dikun/frogger-game/app.js (hunk 5 above)
Comment (YuliiaDikun) [reply submissions/dikun/frogger-game/app.js:105]: but I think I can't use the new keyword...(following "Requirements re OOP:" last point)

Comment (OleksiyRudenko) [reply submissions/dikun/frogger-game/app.js:105]: Classes defined with `prototype`-based syntax are instantiated with `new` keyword. It is way older than the `class` syntax.

The referred requirement is about class **extension**.

File: submissions/dikun/frogger-game/app.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+const SPRITE = {
+	enemy: 'images/enemy-bug.png',
+	player: 'images/char-boy.png'
+};
+
+const SPEED = () => ENEMY_SPEED.min + Math.floor(Math.random() * ENEMY_SPEED.max);
+const OVERLAP_AMOUNT = 70;
+const BORDER = 20;
+
+const Enemy = function (x, y, SPEED, player) {
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:34]: Function parameters (e.g. `SPEED`) are not constants anymore. We may send constants to functions, but the function itself doesn't really care where the parameter value comes from.

**Comment by YuliiaDikun** [comment]
I return to the original inheritance pattern and use a `new` keyword. Also, change my const to uppercase and crash to OVERLAP_AMOUNT. Please, look at the changes! Thank you for all your remarks and comments.

File: submissions/dikun/frogger-game/app.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+};
+
+Player.prototype.render = function () {
+	ctx.drawImage(Resources.get(this.SPRITE), this.x, this.y);
+};
+
+const player = new Player(PLAYER_START.x, PLAYER_START.y);
+
+const allEnemies = [CELL.height - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]
+				   .map(value => {return new Enemy(GAME_FIELD.left, value, SPEED, player);});
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:100]: (1) As a peer developer I am assigned a task to make bugs elegantly flow in from outside. As gracefully as they leave it. Intuition says I need only change the first argument of Enemy constructor (`GAME_FIELD.left` gives a good hint on it). If I go and change the constant value, probably something else will break.
What constant can I subtract from the field left border coord to achieve this? Can you please do this for me?

(2) Arrow functions have a nice feature. If its body has a single statement and it is `return` then function body braces and `return` keyword can be omitted.
```
.map(param => { return param * 2; })

.map(param => param * 2)
```
This adds to code readability.

(3) `value` is to generic word. What this parameter represents in the game domain/world?

File: submissions/dikun/frogger-game/app.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+const GAME_FIELD = {
+	top: 0,
+	right: CELL.width * 5,
+	bottom: CELL.height * 5,
+	left: 0
+};
+
+const PLAYER_START = {
+	x: 202,
+	y: 400
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:17]: I feel these can be calculated.

File: submissions/dikun/frogger-game/app.js [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+	}		
+};
+
+Player.prototype.render = function () {
+	ctx.drawImage(Resources.get(this.SPRITE), this.x, this.y);
+};
+
+const player = new Player(PLAYER_START.x, PLAYER_START.y);
+
+const allEnemies = [CELL.height - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:99]: How can we reduce repetition here? Ideally I'd like to add a row number and call it a day, if I need to add yet another bug.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@YuliiaDikun great improvement to code comprehension!
Let's polish it.

**Comment by YuliiaDikun** [comment]
Please, see my changes. Thank you for your time!

File: submissions/dikun/frogger-game/app.js [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+function createEnemyPerLine (line, numberOfEnemies) {	
+	for (let i = 0; i < numberOfEnemies; i++) {		
+		allEnemies.push(new Enemy((GAME_FIELD.left - CELL.width), LINE[line], speed, player));		
+	}
+}
+
+createEnemyPerLine(1, 1);
+createEnemyPerLine(2, 2);
+createEnemyPerLine(3, 1);
```
Comment (OleksiyRudenko) [inline submissions/dikun/frogger-game/app.js:115]: Mapping was much cleaner and readable.
Now we have longer code and also repetition.
What if I want to add 20 more rows? How can I avoid adding 20 more lines of code?
Take a step back and think how could you optimize `[CELL.height - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]`. There is a pattern there. You must see it better if the line of code modified as follows: `[CELL.height*1 - BORDER, CELL.height*2 - BORDER, CELL.height*3 - BORDER]`.

Think of chained mapping also look into what parameters a callback for `map` receives and you could leverage.

Examples:

```
[1, 5, 7, 3].map(heightMeters => heightMeters * 100).map((heightCentimeters, numberOfWalls) => buildHouseFragment(heightCentimeters, numberOfWalls + 1));
```

Of course the above is optimizable. We can just multiply source data by 100 when calling `buildHouseFragment`. This is specifically to make you start thinking of what numbers like 1, 2, 3 in your original fragment mean.

**Review by OleksiyRudenko — CHANGES_REQUESTED** [review]
@YuliiaDikun nice try.
We need to make a step back and look into another direction.

File: submissions/dikun/frogger-game/app.js (hunk 11 above)
Comment (OleksiyRudenko) [reply submissions/dikun/frogger-game/app.js:115]: Now I can **read** that initial arrays contains lines. That exaplains a lot to me as a peer developer and now I clearly understand what to do if I want to add yet another enemy to run across line 2 for example.
Although I'd suggest rename `line` as `rowNumber` or `laneNumber` or `trackNumber`

File: submissions/dikun/frogger-game/app.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+const GAME_FIELD = {
+	top: 0,
+	right: CELL.width * 5,
+	bottom: CELL.height * 5,
+	left: 0
+};
+
+const PLAYER_START = {
+	x: CELL.width * 2.02,
+	y: CELL.height * 4.7
```
[the remark this answers is not in the document]
Comment (YuliiaDikun) [reply submissions/dikun/frogger-game/app.js]: Fixed.

File: submissions/dikun/frogger-game/app.js (hunk 11 above)
Comment (YuliiaDikun) [reply submissions/dikun/frogger-game/app.js]: I spent a few hours trying to understand how I can add a new bug in line 2. OMG. Thank you a lot for all your remarks. Hope it is fixed now too.

**Comment by YuliiaDikun** [comment]
Please. see my changes. Thanks!

**Review by OleksiyRudenko — APPROVED** [review]
@YuliiaDikun excellent job!

**Pull request merged**